File: katas/csharp/SOLID/ocp/ocp-bad.cs
using System;

// Violación del OCP: Debemos modificar el código existente para agregar nuevos animales
// ❌ Problema: Para agregar un nuevo animal, debemos modificar la clase Communication

public class Dog
{
  public string MakeSound()
  {
    return "woof woof";
  }
}

public class Cat
{
  public string MakeSound()
  {
    return "meow meow";
  }
}

public class Fox
{
  public string MakeSound()
  {
    return "ring-ding-ding-ding-dingeringeding";
  }
}

// ❌ Esta clase debe modificarse cada vez que agregamos un nuevo animal
public class Communication
{
  public string Communicate(object animal)
  {
    // ❌ Cadena de if/else que debe modificarse para cada nuevo animal
    if (animal is Dog dog)
    {
      return dog.MakeSound();
    }
    else if (animal is Cat cat)
    {
      return cat.MakeSound();
    }
    else if (animal is Fox fox)
    {
      return fox.MakeSound();
    }
    else
    {
      throw new Exception("Animal desconocido");
    }
  }
}

// Uso
class Program
{
  static void Main()
  {
    var communication = new Communication();
    var dog = new Dog();
    var cat = new Cat();
    var fox = new Fox();

    Console.WriteLine(communication.Communicate(dog)); // "woof woof"
    Console.WriteLine(communication.Communicate(cat)); // "meow meow"
    Console.WriteLine(communication.Communicate(fox)); // "ring-ding-ding-ding-dingeringeding"

    // ❌ Problema: Para agregar Cow, debemos:
    // 1. Crear la clase Cow
    // 2. Modificar el tipo de parámetro en Communicate() ❌
    // 3. Agregar una nueva rama if/else ❌
  }
}

File: katas/python/SOLID/ocp/ocp_bad.py
# Violación del OCP: Debemos modificar el código existente para agregar nuevos animales
# ❌ Problema: Para agregar un nuevo animal, debemos modificar la clase Communication


class Dog:
    def make_sound(self) -> str:
        return "woof woof"


class Cat:
    def make_sound(self) -> str:
        return "meow meow"


class Fox:
    def make_sound(self) -> str:
        return "ring-ding-ding-ding-dingeringeding"


# ❌ Esta clase debe modificarse cada vez que agregamos un nuevo animal
class Communication:
    def communicate(self, animal) -> str:
        # ❌ Cadena de if/else que debe modificarse para cada nuevo animal
        if isinstance(animal, Dog):
            return animal.make_sound()
        elif isinstance(animal, Cat):
            return animal.make_sound()
        elif isinstance(animal, Fox):
            return animal.make_sound()
        else:
            raise Exception("Animal desconocido")


if __name__ == "__main__":
    # Uso
    communication = Communication()
    dog = Dog()
    cat = Cat()
    fox = Fox()

    print(communication.communicate(dog))  # "woof woof"
    print(communication.communicate(cat))  # "meow meow"
    # "ring-ding-ding-ding-dingeringeding"
    print(communication.communicate(fox))

    # ❌ Problema: Para agregar Cow, debemos:
    # 1. Crear la clase Cow
    # 2. Modificar el tipo de parámetro en communicate() ❌
    # 3. Agregar una nueva rama if/else ❌
